Add support for VolumeAttributeClass (#149)
strimzi/proposals: 095-add-support-volumeattributesclassname.md

# Add `volumeAttributesClassName` to the storage configuration

> Add the ability to set `volumeAttributesClassName` in the configuration for `PersistentClaimStorage`.

Reviewer: Someone reading could not understand what `PersistentVolumeClaim` is here. Of course, it's part of the Strimzi api, it's one of the model classes but need to be clarified.

`VolumeAttributesClasses` provide the ability to decouple storage parameters like
IOPS, throughput, fstype or any other cloud specific ones from the `StorageClass`.

## Current situation

It is  not possible to set the `volumeAttributesClassName` using [`PersistentClaimStorage`](https://github.com/strimzi/strimzi-kafka-operator/blob/c1b20f726dddbcd2a070c2eeb14fd30902027aec/api/src/main/java/io/strimzi/api/kafka/model/kafka/PersistentClaimStorage.java).

> This is currently being managed using various annotations

Reviewer: What does this sentence mean exactly?
"This" is referring to what?
"various annotations" is ambiguos.

Author: Looks like I was incorrect here. I was just referring to the description in strimzi/strimzi-kafka-operator#10462 that requested this proposal. I'll update the document.

## Motivation

Kubernetes v1.31 added a new method of configuring storage parameters for `PersistentVolumes` (PV)
using [`VolumeAttributesClass`](https://kubernetes.io/docs/concepts/storage/volume-attributes-classes/) (VAC).

> `PersistentVolumeClaims` (PVC) can then refer to them using `volumeAttribtesClassName` along with their

Reviewer: ```suggestion
`PersistentVolumeClaims` (PVC) can then refer to them using `volumeAttributesClassName` along with their
```

corresponding `storageClassName`. This decouples storage parameters specification from the `StorageClass`
(SC) into the VAC. In a PVC, the `storageClassName` field is immutable, whereas the
`volumeAttributesClassName` isn't. This makes it possible to dynamically reconfigure the PV without
losing data.

## Proposal

To accommodate this change, the `PersistentClaimStorage` (PCS) API needs an additional field `volumeAttributesClass`.
When this field changes in the PCS, the Cluster Operator (CO) can map it to the generated PVC's
`volumeAttributesClassName` and let the CSI Driver take care of provisioning. There shouldn't be
a need to validate the VAC since the parameter names depend on the cloud provider.

The [`external-provisioner`](https://github.com/kubernetes-csi/external-provisioner) `csi-provisioner`
usually checks whether the `driverName` in the VAC is the same as the `provisioner` in the SC.
The operator could check that pre-emptively, but cannot depend on the CSI driver to use that implementation
of a csi-provisioner. The user would be responsible for making sure the VACs and the SCs are configured

> correctly before configuring the PCS.

Reviewer: What exactly happens if the user doesn't make sure for a proper configuration?

Reviewer 2: I guess the error will surface to the Kafka status. An example would help.

Author: The error is posted on the PVC's status. The user is expected to check the PVC if the pod isn't being scheduled properly. Making it appear in Kafka would mean we are making redundant checks, making sure the PVC is in the ready status (which I'm not sure is currently being done or not). Will update the doc with this info

I have added an example at the end for an AWS EBS backed volume.

Reviewer: You are actually putting a WARN example which is not clear to me where it is visible, some logs? somewhere? On the PVC describe? We were wondering if the user should have a feedback on the Kafka custom resource status that something wrong happened.

Reviewer: Usually the user doesn't need to take a look at the operands (like PVCs in this case) or at least they can get details from there if they want but having a central point for warnings/errors which is the `Kafka` custom resource is much better.

Reviewer 3: I would avoid any further processing as the exact errors might differ depending on the driver etc. I think this is pretty advanced stuff and the users using this should be able to monitor their PVCs etc.

Reviewer: but the error should be in the PVC status not depending on the driver, that's my understanding. So why we can't reflect that in the Kafka CR? (like we reflect errors from other "external systems" in the proper custom resource, i.e. Cruise Control errors into the KafkaRebalance).

Reviewer 3: I do not think it is always the case that it will be there. You also don't know how it will be listed there. I'm not sure what other external systems are you talking about. I think if PVC resizing fails for example, it would also not be necessarily shown in the Kafka CR status as it is hard to detect in all cases.

An example configuration could be

```
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaNodePool
metadata:
  name: pool-a
  labels:
    strimzi.io/cluster: my-cluster
spec:
  replicas: 3
  roles:
  - broker
  resources:
    requests:
      cpu: 1
      memory: 512Gi
    limits:
      cpu: 500m
      memory: 256Gi
  storage:
    type: persistent-claim
    size: 16Gi
    class: ebs
    volumeAttributesClass: ebs-fast
    deleteClaim: true
```

which produces the following PVC

```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: <generated-name>
spec:
  accessMode:
  - ReadWriteOnce
  volumeMode: Filesystem
  resources:
    request:
      storage: 16Gi
  storageClassName: ebs
  volumeAttributesClassName: ebs-fast
```

Here the user is expected to have a VAC named ebs-fast which has the same `driverName` as the
`provisioner` of the SC named ebs. They are also expected to have the right parameters for
the corresponding cloud provider's CSI driver.

When a user decides to update the `volumeAttributesClassName` of a PVC, the CSI driver will

> apply the changes. This might result in a cloud provider error being thrown as well. In that

Reviewer: so does the operator report the error within the `Kafka` status? how does the user notice the mistake in order to fix it?

Reviewer 2: Can you provide an example?

Author: The example is at the end. I have copied it from the PVC's status. Will update the doc with this info

<!-- This template is provided as an example with sections you may wish to comment on with respect to your proposal. Add or remove sections as required to best articulate the proposal. -->

> 

Reviewer 3: I guess you can delete the comment and the empty line.

# Add `volumeAttributesClassName` to the storage configuration

Add the ability to set `volumeAttributesClassName` in the configuration for `PersistentClaimStorage`.
`VolumeAttributesClasses` provide the ability to decouple storage parameters like
IOPS, throughput, fstype or any other cloud specific ones from the `StorageClass`.

> 
> 

Reviewer 3: ```suggestion

```

## Current situation

It is  not possible to set the `volumeAttributesClassName` using [`PersistentClaimStorage`](https://github.com/strimzi/strimzi-kafka-operator/blob/c1b20f726dddbcd2a070c2eeb14fd30902027aec/api/src/main/java/io/strimzi/api/kafka/model/kafka/PersistentClaimStorage.java).
This is currently being managed using various annotations

## Motivation

> Kubernetes v1.31 added a new method of configuring storage parameters for `PersistentVolumes` (PV)

Reviewer 3: What happens on Kube 1.31 while this feature is disabled or on older Kubernetes versions:
* When the field is not set -> I expect nothing will happen and everything will work as before?
* When the field is set -> Does it throw an error? Is it just silently ignored?

Author: > When the field is not set -> I expect nothing will happen and everything will work as before?

Yes, it just functions like before.

> When the field is set -> Does it throw an error? Is it just silently ignored?

It ignores the field. Looks like when I dump the yaml, the PVC doesn't even have that field.

To accommodate this change, the `PersistentClaimStorage` (PCS) API needs an additional field `volumeAttributesClassName`.
When this field changes in the PCS, the Cluster Operator (CO) can map it to the generated PVC and let
the CSI Driver take care of provisioning. There shouldn't be a need to validate the VAC since the
parameter names depend on the cloud provider.

The [`external-provisioner`](https://github.com/kubernetes-csi/external-provisioner) `csi-provisioner`
usually checks whether the `driverName` in the VAC is the same as the `provisioner` in the SC.
The operator could check that pre-emptively, but cannot depend on the CSI driver to use that implementation
of a csi-provisioner. The user would be responsible for making sure the VACs and the SCs are configured
correctly before configuring the PCS.

An example configuration could be

```
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaNodePool
metadata:
  name: pool-a
  labels:
    strimzi.io/cluster: my-cluster
spec:
  replicas: 3
  roles:
  - broker
  resources:
    requests:
      cpu: 1
      memory: 512Gi
    limits:
      cpu: 500m
      memory: 256Gi
  storage:
    type: persistent-claim
    size: 16Gi
    class: ebs

>     volumeAttributesClassName: ebs-fast

Reviewer 3: Givin Storage Class name is just `class`, I wonder if we should name this field `volumeAttributesClass` only (without the `Name` at the end).

Author: Sure will change it :+1: 

I was thinking might be easier to correlate with a PVC if the fields had the same name. But since the `class` field is already not 1-1, I can agree as well.

# Add `volumeAttributesClassName` to the storage configuration

Volumes in Strimzi are configured using the `PersistentClaimStorage` model. This is part
of the Strimzi API's `Kafka` and `KafkaNodePool` among others. This proposal goes over the
ability to set `volumeAttributesClassName` in the configuration for `PersistentClaimStorage`.

> `VolumeAttributesClasses` provide the ability to decouple storage parameters like

Reviewer: ```suggestion
`VolumeAttributesClasse` provides the ability to decouple storage parameters like
```

# Add `volumeAttributesClassName` to the storage configuration

Volumes in Strimzi are configured using the `PersistentClaimStorage` model. This is part of the Strimzi API's `Kafka` and `KafkaNodePool` among others. This proposal goes over the ability to set `volumeAttributesClassName` in the configuration for `PersistentClaimStorage`. `VolumeAttributesClass` provides the ability to decouple storage parameters like IOPS, throughput, fstype or any other cloud specific ones from the `StorageClass`.

## Current situation

It is  not possible to set the `volumeAttributesClassName` using [`PersistentClaimStorage`](https://github.com/strimzi/strimzi-kafka-operator/blob/c1b20f726dddbcd2a070c2eeb14fd30902027aec/api/src/main/java/io/strimzi/api/kafka/model/kafka/PersistentClaimStorage.java). To change any configuration parameters of the physical volume backing a `PersistentVolumeClaim`, users have to rely on the `StorageClass`'s `spec.parameters` field. This couples the parameters and the class.

## Motivation

> Kubernetes v1.31 added a new method of configuring storage parameters for `PersistentVolumes` (PV) using [`VolumeAttributesClass`](https://kubernetes.io/docs/concepts/storage/volume-attributes-classes/) (VAC). `PersistentVolumeClaims` (PVC) can then refer to them using `volumeAttributesClassName` along with their corresponding `storageClassName`. This decouples storage parameters specification from the `StorageClass` (SC) into the VAC. In a PVC, the `storageClassName` field is immutable, whereas the `volumeAttributesClassName` isn't. This makes it possible to dynamically reconfigure the PV without losing data.

Author: The feature itself is fairly breakage free. If the feature gate is enabled and the `volumeAttributeClassName` field is present on the PVC, it is completely ignored when applying.

Otherwise, it is just applied and the CSI driver takes over for any validations.

> Kubernetes v1.31 added a new method of configuring storage parameters for `PersistentVolumes` (PV) using [`VolumeAttributesClass`](https://kubernetes.io/docs/concepts/storage/volume-attributes-classes/) (VAC). `PersistentVolumeClaims` (PVC) can then refer to them using `volumeAttributesClassName` along with their corresponding `storageClassName`. This decouples storage parameters specification from the `StorageClass` (SC) into the VAC. In a PVC, the `storageClassName` field is immutable, whereas the `volumeAttributesClassName` isn't. This makes it possible to dynamically reconfigure the PV without losing data.

Author: > Should we note that VolumeAttributesClass is cluster-wide resource like StorageClass?

I can add that info

> Kubernetes v1.31 added a new method of configuring storage parameters for `PersistentVolumes` (PV) using [`VolumeAttributesClass`](https://kubernetes.io/docs/concepts/storage/volume-attributes-classes/) (VAC). `PersistentVolumeClaims` (PVC) can then refer to them using `volumeAttributesClassName` along with their corresponding `storageClassName`. This decouples storage parameters specification from the `StorageClass` (SC) into the VAC. In a PVC, the `storageClassName` field is immutable, whereas the `volumeAttributesClassName` isn't. This makes it possible to dynamically reconfigure the PV without losing data.

Author: Before I push anything, does pushing dismiss any approvals?

> Kubernetes v1.31 added a new method of configuring storage parameters for `PersistentVolumes` (PV) using [`VolumeAttributesClass`](https://kubernetes.io/docs/concepts/storage/volume-attributes-classes/) (VAC). `PersistentVolumeClaims` (PVC) can then refer to them using `volumeAttributesClassName` along with their corresponding `storageClassName`. This decouples storage parameters specification from the `StorageClass` (SC) into the VAC. In a PVC, the `storageClassName` field is immutable, whereas the `volumeAttributesClassName` isn't. This makes it possible to dynamically reconfigure the PV without losing data.

Reviewer 3: That depends on the amount of changes you do. Explaining tht `VolumeAttributesClass` is cluster wide resource would not cause any issues I guess.

## Proposal

To accommodate this change, the `PersistentClaimStorage` (PCS) API needs an additional field `volumeAttributesClass`. When this field changes in the PCS, the Cluster Operator (CO) can map it to the generated PVC's `volumeAttributesClassName` and let the CSI Driver take care of provisioning. There shouldn't be a need to validate the VAC since the parameter names depend on the cloud provider.

> 

Author: If the error is apparent, e.g. the VAC is incorrect, the pods won't schedule and I think it is safe to expect the users will have alerting for deployment / replicaset mismatch. There are metrics for those. As for ones that don't immediately break the backing storage, there should be kubelet metrics for the volumes as well that say what the state of a PVC is. imo a user should be responsible for any misconfiguration with the CSI side of configuration.

There could be a generic check for if the volume status != failed or something. But that should happen across the controller, regardless of if there is or isn't a VAC. It is out of scope for this proposal.

> To accommodate this change, the `PersistentClaimStorage` (PCS) API needs an additional field `volumeAttributesClass`. When this field changes in the PCS, the Cluster Operator (CO) can map it to the generated PVC's `volumeAttributesClassName` and let the CSI Driver take care of provisioning. There shouldn't be a need to validate the VAC since the parameter names depend on the cloud provider.

Reviewer: I would highlight it's a String field.
